### PR TITLE
[WIP] Only wait to reconnect if there have been 3 consecutive connection failures

### PR DESCRIPTION
**Pull request by hnrkndrssn**
…when accepting a TCP client connection and include link to Tentacle troubleshooting guide. 

Relates to: OctopusDeploy/Issues#5022

### CODE REVIEW

File: source/Halibut/Transport/SecureListener.cs [hunk 1]
```diff
@@ -111,9 +113,18 @@ async Task Accept()
...
+                        log.WriteException(EventType.Error, $"Error accepting TCP client", ex);
+                        // Slow down the logs in case an exception is immediately encountered after 3 failed AcceptTcpClientAsync calls
+                        if (numberOfFailedAttemptsInRow >= 3)
+                        {
+                            var millisecondsTimeout = Math.Max(0, Math.Min(numberOfFailedAttemptsInRow - 3, 100)) * 10;
+                            log.Write(
+                                EventType.Diagnostic,
+                                $"Connection has failed {numberOfFailedAttemptsInRow} times. Waiting {millisecondsTimeout}ms before retrying. For a detailed troubleshooting guide go to https://g.octopushq.com/TentacleTroubleshooting"
+                            );
+                            Thread.Sleep(millisecondsTimeout);
```
Comment (matt-richardson) [inline source/Halibut/Transport/SecureListener.cs]: Bit of a drive by, but would `Task.Delay()` be more appropriate here?

File: source/Halibut/Transport/SecureListener.cs [hunk 2]
```diff
@@ -111,9 +113,18 @@ async Task Accept()
...
-                        Thread.Sleep(1000);
+                        numberOfFailedAttemptsInRow++;
+                        log.WriteException(EventType.Error, $"Error accepting TCP client", ex);
+                        // Slow down the logs in case an exception is immediately encountered after 3 failed AcceptTcpClientAsync calls
+                        if (numberOfFailedAttemptsInRow >= 3)
+                        {
+                            var millisecondsTimeout = Math.Max(0, Math.Min(numberOfFailedAttemptsInRow - 3, 100)) * 10;
+                            log.Write(
+                                EventType.Diagnostic,
+                                $"Connection has failed {numberOfFailedAttemptsInRow} times. Waiting {millisecondsTimeout}ms before retrying. For a detailed troubleshooting guide go to https://g.octopushq.com/TentacleTroubleshooting"
```
Comment (droyad) [inline source/Halibut/Transport/SecureListener.cs]:
```suggestion
                                $"Accepting a connection has failed {numberOfFailedAttemptsInRow} times in a row. Waiting {millisecondsTimeout}ms before attempting to accept another connection. For a detailed troubleshooting guide go to https://g.octopushq.com/TentacleTroubleshooting"
```

File: source/Halibut/Transport/SecureListener.cs [hunk 3]
```diff
@@ -111,9 +113,18 @@ async Task Accept()
...
-                        // Slow down the logs in case an exception is immediately encountered each AcceptTcpClientAsync
-                        Thread.Sleep(1000);
+                        numberOfFailedAttemptsInRow++;
+                        log.WriteException(EventType.Error, $"Error accepting TCP client", ex);
+                        // Slow down the logs in case an exception is immediately encountered after 3 failed AcceptTcpClientAsync calls
+                        if (numberOfFailedAttemptsInRow >= 3)
+                        {
+                            var millisecondsTimeout = Math.Max(0, Math.Min(numberOfFailedAttemptsInRow - 3, 100)) * 10;
+                            log.Write(
+                                EventType.Diagnostic,
```
Comment (droyad) [inline source/Halibut/Transport/SecureListener.cs]:
```suggestion
                                EventType.Error,
```

File: source/Halibut/Transport/SecureListener.cs (hunk 1 above)
Comment (droyad) [reply source/Halibut/Transport/SecureListener.cs]: It's much easier to see where this process is hanging, which was valuable in finding this bug. I'll do up a PR to make this method not async

**Pull request merged**